_episodes/02-keras.md

```diff
@@ -0,0 +1,514 @@
+to use for those.
+Furthermore the training we need to specify how many times to show the training samples.
+
+Once more keras gives us plenty of choices all have their own pro's and cons,
+but for now let us go with the widely used Adam optimizer.
+Adam has a number of parameters, but the default values work well for most problems.
+So we will use it with its default parameters.
+
+Combining this with the loss function we decided on earlier we can now compile the
+model using `model.compile`.
+Compiling the model prepares it to start the training.
+
+Training the model is done using the `fit` method, it takes the input data and
+target data as inputs and it has several other parameters for certain options
+of the training.
+Here we only set a different number of `epochs`.
+One training epoch means that every sample in the training data has been shown
+to the neural network and used to update its parameters.
+
+~~~
+model.compile(optimizer=keras.optimizers.Adam(), loss=keras.losses.CategoricalCrossentropy())
+history = model.fit(training_data, target, epochs=100)
+~~~
+{:.language-python}
+
+The fit method returns a history object that has a history attribute with the training loss and
+potentially other metrics per training epoch.
+It can be very insightful to plot the training loss to see how the training progresses.
+Using seaborn we can do this as follow:
+~~~
+sns.lineplot(x=history.epoch, y=history.history['loss'])
+~~~
+{:.language-python}
+
+## 7. Tune hyperparameters
+As we discussed before the design and training of a neural network comes with
+many hyper parameter choices.
+We will go into more depth of these hyperparameters in later episodes, but
+it is good to play with them a little bit to see how they change and influence the
+model training.
+> ## Change some of the hyperparameters
+>
+> Change one of the hyperparameters of the neural network.
+> You can for instance change the model by adding or removing layers,
+> changing the activation function of the hidden layer.
+> Or you can change the training, for instance by adding a learning
+> rate throught the `learning_rate` parameter of the Adam optimizer.
+>
+> * How does changing these hyperparameters impact the training
+> * What would be a good strategy to find good hyperparameter values?
+{:.challenge}
+
+## 8. Measuring Performance
+Normally we would now try to assess the quality of the trained model by testing it
+on some data that was not used for training.
+This is often done by keeping a portion of the dataset seperate from the training data,
+when using that to measure performance such a set is called a test set.
+In this episode we did not prepare such a test set, however, and this will be covered
+in later epidsodes.
+
+## 9. 'predict'
+Now that we have a training neural network, we can use it to predict new samples
+of penguin using the `predict` function.
+This will use the network to predict the outputs we trained.
```

Comment:
I think it would be good to leave out a test set at the beginning to do prediction on and a final test of our model. The next episode can then add the step of also holding out a validation set for hyperparameter tuning.